redmine: 4.2.4 -> 4.2.5

felixsinger opened the pull request:
###### Description of changes

Add an update script and update Redmine to version 4.2.5.

Change log: https://www.redmine.org/projects/redmine/wiki/Changelog_4_2

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

felixsinger committed:
redmine: 4.2.4 -> 4.2.5

Update Redmine to version 4.2.5.

Change log: https://www.redmine.org/projects/redmine/wiki/Changelog_4_2

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger commented:
How do I update the files Gemfile.lock and gemset.nix?

aanderse commented:
how to update `redmine`:
```shell
cd ~/nixpkgs/pkgs/applications/version-management/redmine/
# update default.nix to new version number/hash
rm Gemfile Gemfile.lock gemset.nix
# copy Gemfile from new redmine version
# manually modify the database lines in Gemfile (diff the two files, it's obvious)
nix-shell -p bundix --run "bundix -l"
```

felixsinger committed:
redmine: Add update script

The following steps need to be done before running the script:

  1. Copy Gemfile from new Redmine version to the Redmine folder in
     Nixpkgs
  2. Manually modify the database lines in Gemfile and just include the
     gems for the supported databases (diff the two files).

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger commented:
Thanks :) I added a script to do that. So it's somewhat documented.

aanderse commented:
Awesome! Any chance you want to take over as maintainer of this package+module?

felixsinger commented:
> Awesome! Any chance you want to take over as maintainer of this package+module?

Sounds like you are planning to leave the maintainership? I wouldn't like to do that alone, since I still consider myself as a NixOS beginner.

So I agree if you also stay as a maintainer. I am happy to help :)

felixsinger commented on pkgs/applications/version-management/redmine/Gemfile.lock:
```diff
@@ -206,7 +204,7 @@ GEM
 
 PLATFORMS
-  ruby
+  x86_64-linux
```
I guess this line needs to be reverted?

aanderse commented on pkgs/applications/version-management/redmine/gemset.nix:
```diff
@@ -405,12 +395,12 @@
     platforms = [];
     source = {
       remotes = ["https://rubygems.org"];
-      sha256 = "1vrn31385ix5k9b0yalnlzv360isv6dincbcvi8psllnwz4sjxj9";
```
Are you seeing this too?
```
error: hash mismatch in fixed-output derivation '/nix/store/1h2cfzdd32rhvnlwkq16z84rc525rri9-nokogiri-1.11.7.gem.drv':
         specified: sha256-etl0Hnov7h/7SksuILAOh5ksnv2Wn1V8o7g/smU7m/w=
            got:    sha256-SXapyeeWUn1R3GwxG5vZOgIz9qeWKg9WmqXHgkYYNu8=

```

felixsinger replied:
Just tried to build. Yes, I'm getting this too.

felixsinger replied:
Rerunning bundix doesn't help.

aanderse replied:
So why is `bundix` pulling the wrong hash... :thinking: I'll have to get back to this at some point. If you find anything please ping.

felixsinger replied:
I think I found the issue... For some reason, bundix doesn't update the version string of nokogiri. It's still 1.11.7.

felixsinger replied:
The output of bundix says 1.11.7 is downloaded. So my guess is wrong.
```
path is '/nix/store/mix7sp02x07130wdc6rfyma3l9ndb0dc-nokogiri-1.11.7-x86_64-linux.gem'
1z4v7djv4gxqldy5b7wnzng2r6c71sq20bjb9bxizvigg8g79nbs => nokogiri-1.11.7-x86_64-linux.gem
```

felixsinger replied:
Now I found the issue :)

This is what bundix downloads:
```
$ nix-prefetch-url https://rubygems.org/downloads/nokogiri-1.11.7-x86_64-linux.gem
[3.4 MiB DL]
path is '/nix/store/mix7sp02x07130wdc6rfyma3l9ndb0dc-nokogiri-1.11.7-x86_64-linux.gem'
1z4v7djv4gxqldy5b7wnzng2r6c71sq20bjb9bxizvigg8g79nbs
```

This is what Nix downloads:
```
$ nix-prefetch-url https://rubygems.org/downloads/nokogiri-1.11.7.gem
[8.9 MiB DL]
path is '/nix/store/a35hgacdgwh9a52mx6ssf0ijb60axkcn-nokogiri-1.11.7.gem'
1vrn31385ix5k9b0yalnlzv360isv6dincbcvi8psllnwz4sjxj9
```

Maybe this is related to the line in Gemfile.lock I marked before? But this file is also generated by bundix...
```
PLATFORMS
-  ruby
+  x86_64-linux
```

felixsinger replied:
I changed according lines manually now. This is not the right place to discuss this.

aanderse replied:
So we're going to run into this issue again next time we update `redmine`, right? Should we talk to upstream `bundix` and see what we need to adjust?

felixsinger replied:
Most likely, yes. Let's reach out to the bundix developers.

aanderse commented:
@ofborg test redmine.mysql redmine.pgsql

aanderse reviewed (approved):
LGTM :+1:

Good to merge?

felixsinger commented:
> LGTM +1
> 
> Good to merge?

Since the test for aarch64 failed, should we mark it as broken?

felixsinger reviewed (commented):
I can't block my own pull request, if I want to. -.-' * rant *

However, looking at an older version update, this one went well. aarch64 didn't fail in the tests. So I think we should try to fix that problem here. https://github.com/NixOS/nixpkgs/pull/161895/checks?check_run_id=5344121453

felixsinger commented:
@GrahamcOfBorg test redmine.mysql redmine.pgsql

felixsinger commented:
Looks like my last change fixed the aarch64 issue! :)

aanderse commented:
There are a number of manual steps you took in this PR, right? Can we document this somehow? The next time someone updates this package it will be very confusing if we don't.

felixsinger commented:
> There are a number of manual steps you took in this PR, right? Can we document this somehow? The next time someone updates this package it will be very confusing if we don't.

In addition to your steps, I did the following steps after yours, but not sure why they are necessary here. It looks like it worked for you a number of times.
* Remove the extension `-x86_64-linux` from the `nokogiri` gem listed in the file `Gemfile.lock`
* In the `PLATFORMS` section of the file `Gemfile.lock`, replace `x86_64-linux` with `ruby`
* Go to https://rubygems.org/gems/nokogiri/versions and download the needed package of the nokogiri gem manually using `nix-prefetch-url` (so that the hash is printed). Important note: Look for the version (here 1.11.7) and then make sure to use that package which isn't specific for any platform. Then, replace the hash of the nokogiri gem in `gemset.nix` with that one which is printed by `nix-prefetch-url`.

felixsinger commented:
I would say let's keep the steps here for now and see how the next update goes. Maybe this was just a random issue.

Just wondering, can you reproduce this issue? Maybe something with my setup is wrong? I used your steps as you described. I executed them manually and with that update script.

felixsinger commented:
@aanderse Anything left to do here?

aanderse commented:
> Just wondering, can you reproduce this issue?

I wanted to confirm this before merging but unfortunately I won't have the time in the next few weeks. Let's merge this and see what happens next time, as you suggest.

Thanks!
